src/pages/SignIn/index.js

```diff
@@ -1,19 +1,55 @@
-import React, { useState } from 'react';
+import React, { useState, useEffect, useRef } from 'react';
 
 import Button from '@material-ui/core/Button';
 import Checkbox from '@material-ui/core/Checkbox';
 import FormControlLabel from '@material-ui/core/FormControlLabel';
+import { useLocation } from 'react-router-dom';
+
+import queryString from '@/package/queryString';
 
 import antiqueUniverseLogo from '../../assets/antique-universe-logo.png';
 
+import KaKaoOauthStrategy from './KaKaoOauthStrategy';
+import OauthStrategy from './OauthStrategy';
+
 import './signin.css';
 
+const KAKAO_VENDOR = 'KAKAO';
+
 const SignIn = () => {
+  const location = useLocation();
   const [staySigninState, setStaySigninState] = useState(false);
+  const oauthStrategy = useRef(new OauthStrategy());
+  const { code } = queryString.parse(location.search);
+  const vendor = window.localStorage.getItem('vendor');
+
+  useEffect(() => {
+    if (vendor === KAKAO_VENDOR) {
+      oauthStrategy.current = new KaKaoOauthStrategy();
+    } else {
+      oauthStrategy.current = new OauthStrategy();
+    }
+  }, [vendor]);
 
-  const clickSigninBtn = () => alert('Kakao signin');
+  useEffect(() => {
+    if (code) {
+      (async () => {
+        const token = await oauthStrategy.current.requestToken(code);
+        console.log(token);
+      })();
+    }
+  }, [code]);
 
-  const clickStaySigninCheckbox = () => setStaySigninState(!staySigninState);
+  const clickSigninBtn = () => {
+    oauthStrategy.current.requestCode();
+  };
+  const clickKakaoSigninBtn = () => {
+    window.localStorage.setItem('vendor', KAKAO_VENDOR);
+    clickSigninBtn();
```

Comment:
뭔가 전역 `oauthStrategy`를 특정 경우에 변경하면서 사용하는 것 같은데, 다음과 같이 각 함수들이 전략을 주입받도록 하는 것은 어떤가요?

```suggestion
    clickSigninBtn(new KaKaoOauthStrategy());
```

```
함수를 순수하게 만들면 좋을 것 같습니다.
```

Reply:
오오 네 이거는 `vendor` 랑 의존성이 얽혀있어서 위 코드에서 조금 바꿔서 순수함수를 가지도록 변경하겠습니다